src/main/java/de/danielbechler/diff/selector/CollectionItemElementSelector.java

```diff
@@ -71,6 +94,6 @@ public boolean equals(final Object o)
 	@Override
 	public int hashCode()
 	{
-		return item != null ? item.hashCode() : 0;
+		return item != null ? identityStrategy.hashCode(item) : 0;
```

Comment:
It's confusing, prone to errors and unnecessary. Why would you want to keep it when it serves no purpose but has high potential for confusing the user?

My very first test of the feature resulted in a pretty long debugging session. Here is the code without the fix:

```groovy
/*
 * Copyright 2015 Daniel Bechler
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package de.danielbechler.diff.identity

import de.danielbechler.diff.ObjectDiffer
import de.danielbechler.diff.ObjectDifferBuilder
import de.danielbechler.diff.node.DiffNode
import de.danielbechler.diff.node.PrintingVisitor
import de.danielbechler.diff.path.NodePath
import de.danielbechler.diff.selector.CollectionItemElementSelector
import de.danielbechler.util.Objects
import groovy.transform.EqualsAndHashCode
import spock.lang.Specification

class CustomIdentitiyStrategyAT extends Specification {

	def 'test'() {
		Pancake base = new Pancake(name: 'Bacon Pancake', toppings: [
				new Topping(id: 1, name: 'Bacon')
		])
		Pancake working = new Pancake(name: 'Bacon Pancake', toppings: [
				new Topping(id: 3, name: 'Bacon'),
				new Topping(id: 2, name: 'More Bacon')
		])

		given:
		  IdentityStrategy identityStrategy = new ToppingNameIdentityStrategy()
		  ObjectDiffer objectDiffer = ObjectDifferBuilder.startBuilding()
				  .identity().ofType(Topping).toUse(identityStrategy).and()
				  .build()
		when:
		  DiffNode node = objectDiffer.compare(working, base)
		  node.visit(new PrintingVisitor(working, base))
		then:
		  def bacon = new Topping(id: 3, name: 'Bacon')
		  //  ^^^^^ I deliberately clone the bacon topping to make sure comparison via == returns false
		  def baconPath = NodePath.startBuilding()
				  .propertyName('toppings')
				  .element(new CollectionItemElementSelector(bacon, identityStrategy))
				  .build()
		  def baconNode = node.getChild(baconPath)
		  baconNode != null
		  baconNode.canonicalGet(working) == new Topping(id: 1, name: 'Bacon')
	}

	class ToppingNameIdentityStrategy implements IdentityStrategy {
		@Override
		boolean equals(Object thisObject, Object thatObject) {
			if (thisObject instanceof Topping && thatObject instanceof Topping) {
				final Topping thisTopping = (Topping) thisObject
				final Topping thatTopping = (Topping) thatObject
				return Objects.isEqual(thisTopping.name, thatTopping.name)
			}
			return false
		}

		@Override
		int hashCode(Object _this) {
			return _this.hashCode();
		}
	}

	static class Pancake {
		String name
		List<Topping> toppings = []
	}

	@EqualsAndHashCode(includes = ['id'])
	static class Topping {
		int id
		String name

		@Override
		public String toString() {
			final StringBuilder sb = new StringBuilder("Topping{");
			sb.append("id=").append(id);
			sb.append(", name='").append(name).append('\'');
			sb.append('}');
			return sb.toString();
		}
	}
}
```

The code results in the following error:

```
Condition not satisfied:

baconNode != null
|         |
null      false

	at de.danielbechler.diff.identity.CustomIdentitiyStrategyAT.test(CustomIdentitiyStrategyAT.groovy:56)
```

Can you spot the problem?